lib/translator/unmap.cc

```diff
@@ -0,0 +1,92 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "unmap.h"
+
+namespace translator {
+
+// Section 5.6
+// https://www.nvmexpress.org/wp-content/uploads/NVM-Express-SCSI-Translation-Reference-1_1-Gold.pdf
+StatusCode UnmapToNvme(absl::Span<const uint8_t> scsi_cmd,
+                       absl::Span<const uint8_t> buffer_out,
+                       nvme::GenericQueueEntryCmd& nvme_cmd,
+                       Allocation& allocation, uint32_t nsid) {
+  scsi::UnmapCommand unmap_cmd;
+  if (!ReadValue(scsi_cmd, unmap_cmd)) {
+    DebugLog("Malformed unmap command");
+    return StatusCode::kFailure;
+  }
+  if (unmap_cmd.param_list_length < sizeof(scsi::UnmapParamList)) {
+    DebugLog("Insufficient unmap parameter list length");
+    return StatusCode::kFailure;
+  }
+  if (unmap_cmd.anchor == 1) {
+    DebugLog("Unsupported unmap anchor request");
+    return StatusCode::kNoTranslation;
+  }
+
+  // Copy unmap parameter list into memory
+  scsi::UnmapParamList param_list;
+  if (!ReadValue(buffer_out, param_list)) {
+    DebugLog("Malformed unmap parameter list");
+    return StatusCode::kFailure;
+  }
+  buffer_out = buffer_out.subspan(sizeof(scsi::UnmapParamList));
+
+  // Ensure that block descriptor data length & span length align
+  if (buffer_out.size() < param_list.block_desc_data_length) {
+    DebugLog("Block descriptor list length reported longer than buffer");
+    return StatusCode::kFailure;
+  }
+  uint32_t block_descriptor_count =
+      param_list.block_desc_data_length / sizeof(scsi::UnmapBlockDescriptor);
```

Comment:
do we need to worry about divisibility?

Reply:
Good catch, technically a non-divisible block_desc_data_length is a SCSI check condition error. I'll update accordingly